File: csharp/Behavioral/17-Memento/Originator/JobCard.cs
public class JobCard
{
    public string JobId { get; init; } = Guid.NewGuid().ToString()[..8];
    public string Title { get; init; } = string.Empty;
    public string Category { get; init; } = string.Empty;
    private string _status = "Draft";
    private string? _assignee;
    private string? _notes;
    private decimal _cost;

    public string Status => _status;
    public string? Assignee => _assignee;
    public string? Notes => _notes;
    public decimal Cost => _cost;

    public JobCard(string title, string category, decimal cost)
    {
        Title = title;
        Category = category;
        _cost = cost;
    }

    public void Assign(string handyman)
    {
        _assignee = handyman;
        _status = "Assigned";
    }

    public void StartWork() => _status = "InProgress";

    public void PauseWork(string reason)
    {
        _status = "Paused";
        _notes = reason;
    }

    public void Complete()
    {
        _status = "Completed";
        _notes = null;
    }

    public void UpdateCost(decimal newCost) => _cost = newCost;

    public JobCardMemento Save(string label = "") =>
        new JobCardMemento(label, _status, _assignee, _notes, _cost);

    public void Restore(JobCardMemento memento)
    {
        _status = memento.Status;
        _assignee = memento.Assignee;
        _notes = memento.Notes;
        _cost = memento.Cost;
    }

    public override string ToString() =>
        $"Job #{JobId} '{Title}' [{Category}]" +
        $"\n  Status  : {_status}" +
        $"\n  Assignee: {_assignee ?? "none"}" +
        $"\n  Notes   : {_notes ?? "none"}" +
        $"\n  Cost    : N${_cost:N2}";

    public sealed class JobCardMemento
    {
        internal string Status { get; }
        internal string? Assignee { get; }
        internal string? Notes { get; }
        internal decimal Cost { get; }

        public string Label { get; }
        public DateTime SavedAt { get; } = DateTime.Now;

        internal JobCardMemento(
            string label, string status,
            string? assignee, string? notes, decimal cost)
        {
            Label = label;
            Status = status;
            Assignee = assignee;
            Notes = notes;
            Cost = cost;
        }

        public override string ToString() =>
            $"[{SavedAt:HH:mm:ss}] Snapshot: '{Label}' — Status: {Status}";
    }
}

File: typescript/Behavioral/17-Memento/Originator/JobCard.ts
import { JobCardMemento } from "../Memento/JobCardMemento";

export class JobCard {
  readonly jobId: string = Math.random().toString(36).slice(2, 10);

  private _status: string = "Draft";
  private _assignee: string | undefined;
  private _notes: string | undefined;
  private _cost: number;

  get status(): string {
    return this._status;
  }
  get assignee(): string | undefined {
    return this._assignee;
  }
  get notes(): string | undefined {
    return this._notes;
  }
  get cost(): number {
    return this._cost;
  }

  constructor(
    readonly title: string,
    readonly category: string,
    cost: number,
  ) {
    this._cost = cost;
  }

  assign(handyman: string): void {
    this._assignee = handyman;
    this._status = "Assigned";
  }

  startWork(): void {
    this._status = "InProgress";
  }

  pauseWork(reason: string): void {
    this._status = "Paused";
    this._notes = reason;
  }

  complete(): void {
    this._status = "Completed";
    this._notes = undefined;
  }

  updateCost(newCost: number): void {
    this._cost = newCost;
  }

  save(label: string = ""): JobCardMemento {
    return new JobCardMemento(label, {
      status: this._status,
      assignee: this._assignee,
      notes: this._notes,
      cost: this._cost,
    });
  }

  restore(memento: JobCardMemento): void {
    const state = memento.getState();
    this._status = state.status;
    this._assignee = state.assignee;
    this._notes = state.notes;
    this._cost = state.cost;
  }

  toString(): string {
    return [
      `Job #${this.jobId} '${this.title}' [${this.category}]`,
      `  Status  : ${this._status}`,
      `  Assignee: ${this._assignee ?? "none"}`,
      `  Notes   : ${this._notes ?? "none"}`,
      `  Cost    : N$${this._cost.toFixed(2)}`,
    ].join("\n");
  }
}
